rules/run_binary.bzl

```diff
@@ -0,0 +1,105 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+run_binary() build rule implementation.
+
+Runs a binary as a build action. This rule does not require Bash (unlike native.genrule()).
+"""
+
+load("//lib:dicts.bzl", "dicts")
+
+def _make_args(ctx):
+    targets = [ctx.attr.tool]  #+ ctx.attr.srcs
+    resolved = [
+        ctx.expand_location(a, targets) if "$(location" in a else a
+        for a in ctx.attr.args
+    ]
+    result = []
+    for v in resolved:
+        result += ctx.tokenize(v)
+    return result
+
+def _make_envs(ctx):
+    targets = [ctx.attr.tool]  #+ ctx.attr.srcs
+    return {
+        k: ctx.expand_location(v, targets) if "$(location" in v else v
+        for k, v in ctx.attr.env.items()
+    }
+
+def _impl(ctx):
+    tool_inputs, tool_input_mfs = ctx.resolve_tools(tools = [ctx.attr.tool])
+    args = _make_args(ctx)
+    envs = _make_envs(ctx)
+    ctx.actions.run(
+        outputs = ctx.outputs.outs,
+        inputs = depset(direct = ctx.files.srcs, transitive = [tool_inputs]),
+        executable = ctx.executable.tool,
+        arguments = args,
+        mnemonic = "RunBinary",
+        progress_message = "Running tool",
```

Comment:
I don't think this message adds any value.
Maybe don't set it and keep the default message (we could also improve the default message).

Reply:
Done.